cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,299 @@
+        return scale_type{-scale};
+    }
+
+    // perform this operation when constructing with positive scale
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale > 0);
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // perform this operation when constructing with negative scale
+    template <Radix Rad, typename T>
+    constexpr auto left_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale < 0);
+        return val * std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(negate(scale)));
+    }
+
+    // convenience generic shift function
+    template <Radix Rad, typename T>
+    constexpr auto shift(T const& val, scale_type const& scale) -> double {
+        if      (scale == 0) return static_cast<double>(val);
+        else if (scale >  0) return right_shift<Rad>(val, scale);
+        else                 return left_shift <Rad>(val, scale);
+    }
+}
+
+template <typename T>
+constexpr inline auto is_supported_representation_type() -> bool {
+  return std::is_same<T, int32_t>::value;
+}
+
+template <typename T>
+constexpr inline auto is_supported_construction_value_type() -> bool {
+  return std::is_integral      <T>::value
+      || std::is_floating_point<T>::value;
+}
+
+// helper struct for constructing fixed_point when value is already shifted
+template <typename Rep,
+          typename std::enable_if_t<is_supported_representation_type<Rep>()>* = nullptr>
+struct scaled_integer{
+    Rep value;
+    scale_type scale;
+    explicit scaled_integer(Rep v, scale_type s) : value(v), scale(s) {}
+};
+
+// Rep = representative type
+template <typename Rep, Radix Rad>
+class fixed_point {
+
+    Rep        _value;
+    scale_type _scale;
+
+public:
+
+    // CONSTRUCTORS
+    template <typename T,
+              typename std::enable_if_t<is_supported_construction_value_type<T>()
+                                     && is_supported_representation_type<Rep>()>* = nullptr>
+    explicit fixed_point(T const& value, scale_type const& scale) :
+        _value{static_cast<Rep>(detail::shift<Rad>(value, scale))},
+        _scale{scale}
+    {
+    }
+
+    explicit fixed_point(scaled_integer<Rep> s) :
+        _value{s.value},
+        _scale{s.scale}
+    {
+    }
+
+    // DEFAULT CONSTRUCTOR
+    fixed_point() :
+        _value{0},
+        _scale{scale_type{0}}
+    {
+    }
+
+    // EXPLICIT CONVERSION OPERATOR
+    template <typename U,
+              typename std::enable_if_t<is_supported_construction_value_type<U>()>* = nullptr>
+    explicit constexpr operator U() const {
+        return detail::shift<Rad>(static_cast<U>(_value), detail::negate(_scale));
+    }
+
+    auto get() const noexcept {
+        return static_cast<double>(*this);
+    }
+
+    template <typename Rep1, Radix Rad1>
+    fixed_point<Rep1, Rad1>& operator+=(fixed_point<Rep1, Rad1> const& rhs) {
```

Comment:
Shouldn't all the operators be `__host__` `__device__`?

Reply:
Added.